### PR TITLE
Snapshot restore into new PVC

### PR DESCRIPTION
**Pull request by smitaraut**
This PR adds the function to restore a volume snapshot into a new PVC by means of VolumeSource in CreateVolume operation.

**Details of changes:**
- Created wrapper method over copy snapshot REST API
- Enhanced CreateVolume API to take care of snapshot volume source
- Changed the snapshot ID to contain relative path within the snapshot directory
- Changed the file mode of plugin directory from 755 to 644 as per SPBD requirement
- Updated provisioner cluster role to access VolumeSnapshots and VolumeSnapshotContents
- SNAPSHOT.md changes
- Added PVC manifest with volume source in examples
- Added nodeClass parameter to storageclass
- Bumped up snapshotter version to v2.2.0-rc1

**Testing Done:**
- Ran automation bucket
- Snapshot creation/deletion
- Created PVC from snapshot
- PVC creation/deletion

**ID Impact:**
- Manifest for volume to be created from source snapshot
- nodeclass documentation
- min scale version update

### CODE REVIEW

File: driver/cmd/ibm-spectrum-scale-csi/main.go [hunk 1]
```diff
@@ -69,7 +69,7 @@ func handle() {
 
 func createPersistentStorage(persistentStoragePath string) error {
 	if _, err := os.Stat(persistentStoragePath); os.IsNotExist(err) {
-		if err := os.MkdirAll(persistentStoragePath, os.FileMode(0755)); err != nil {
+		if err := os.MkdirAll(persistentStoragePath, os.FileMode(0644)); err != nil {
```
Comment (Jainbrt) [inline driver/cmd/ibm-spectrum-scale-csi/main.go:72]: Is this change related to issue 259 ?

Comment (smitaraut) [reply driver/cmd/ibm-spectrum-scale-csi/main.go:72]: Yes, its related to issue 259

File: driver/csiplugin/controllerserver.go [hunk 2]
```diff
@@ -521,15 +564,116 @@ func (cs *ScaleControllerServer) CreateVolume(ctx context.Context, req *csi.Crea
...
+	if !isFsetLinked {
+		return status.Error(codes.Internal, fmt.Sprintf("Fileset [%v] of source snapshot is not linked", sId.FsetName))
+	}
+
+	isSnapExist, err := conn.CheckIfSnapshotExist(sId.FsName, sId.FsetName, sId.SnapName)
+	if err != nil {
+		return status.Error(codes.Internal, fmt.Sprintf("unable to get snapshot information for [%v]", sId.SnapName))
+	}
+	if !isSnapExist {
+		return status.Error(codes.Internal, fmt.Sprintf("snapshot [%v] does not exist for dileset [%v]", sId.SnapName, sId.FsetName))
```
Comment (deeghuge) [inline driver/csiplugin/controllerserver.go:651]: typo- it should be fileset

File: driver/csiplugin/controllerserver.go [hunk 3]
```diff
@@ -400,6 +413,31 @@ func (cs *ScaleControllerServer) CreateVolume(ctx context.Context, req *csi.Crea
...
+
+		srcSnap := volSrc.GetSnapshot()
+		if srcSnap != nil {
+			snapId := srcSnap.GetSnapshotId()
+			snapIdMembers, err = cs.GetSnapIdMembers(snapId)
+			if err != nil {
+				glog.Errorf("CreateVolume [%s]: Invalid snapshot ID %s [%v]", volName, snapId, err)
+				return nil, err
+			}
+			err = cs.ValidateSnapId(&snapIdMembers, scaleVol, PCid)
```
Comment (deeghuge) [inline driver/csiplugin/controllerserver.go:432]: In case of volume create from snapshot there are duplicate check being added because of ValidateSnapId function, some the check are there after this function.

File: driver/csiplugin/controllerserver.go [hunk 4]
```diff
@@ -899,37 +1039,28 @@ func (cs *ScaleControllerServer) CreateSnapshot(ctx context.Context, req *csi.Cr
...
+		return nil, status.Error(codes.Internal, fmt.Sprintf("Unable to get the snapshot details for [%s]. Error [%v]", snapName, err))
 	}
 
 	if !snapExist {
-		glog.V(5).Infof("CreateSnapshot - creating snapshot [%s] in fileset [%s] under filesystem [%s]", snapName, filesetName, filesystemName)
 		snaperr := conn.CreateSnapshot(filesystemName, filesetName, snapName)
+		glog.Errorf("CreateSnapshot [%s] - Unable to create snapshot. Error [%v]", snapName, snaperr)
 		if snaperr != nil {
-			return nil, status.Error(codes.Internal, fmt.Sprintf("CreateSnapshot - Unable to create snapshot. Error [%v]", snaperr))
+			return nil, status.Error(codes.Internal, fmt.Sprintf("Unable to create snapshot [%s]. Error [%v]", snapName, snaperr))
```
Comment (deeghuge) [inline driver/csiplugin/controllerserver.go:1052]: as per my vscode, error msg must start lower case letter and should not end with fullstop. I try to follow that. I use unable instead of Unable

File: driver/csiplugin/controllerserver.go (hunk 2 above)
Comment (smitaraut) [reply driver/csiplugin/controllerserver.go]: done

File: driver/csiplugin/controllerserver.go (hunk 3 above)
Comment (smitaraut) [reply driver/csiplugin/controllerserver.go]: Moved validateSnapId call below other validations and optimized some validations in the function to avoid repeated checks.

File: driver/csiplugin/controllerserver.go (hunk 4 above)
Comment (smitaraut) [reply driver/csiplugin/controllerserver.go]: Done for instances related to this PR

File: driver/csiplugin/controllerserver.go [hunk 5]
```diff
@@ -900,36 +1043,30 @@ func (cs *ScaleControllerServer) CreateSnapshot(ctx context.Context, req *csi.Cr
...
 	snapExist, err := conn.CheckIfSnapshotExist(filesystemName, filesetName, snapName)
 	if err != nil {
-		return nil, status.Error(codes.Internal, fmt.Sprintf("CreateSnapshot - Unable to get the snapshot details. Error [%v]", err))
+		glog.Errorf("CreateSnapshot [%s] - Unable to get the snapshot details. Error [%v]", snapName, err)
+		return nil, status.Error(codes.Internal, fmt.Sprintf("Unable to get the snapshot details for [%s]. Error [%v]", snapName, err))
 	}
 
 	if !snapExist {
 		snaperr := conn.CreateSnapshot(filesystemName, filesetName, snapName)
+		glog.Errorf("snapshot [%s] - Unable to create snapshot. Error [%v]", snapName, snaperr)
```
Comment (deeghuge) [inline driver/csiplugin/controllerserver.go:1052]: this need to under if snaperr != nil

File: SNAPSHOTS.md [hunk 6]
```diff
@@ -96,11 +96,35 @@ Snapshot should be in "readytouse" state and a corresponding fileset snapshot sh
...
+      resources:
+         requests:
+            storage: 1Gi
+   storageClassName: scfilesetinode
+   dataSource:
+      name: snap1
+      kind: VolumeSnapshot
+      apiGroup: snapshot.storage.k8s.io
+
+    ```
```
Comment (deeghuge) [inline SNAPSHOTS.md:128]: extra ```

Comment (smitaraut) [reply SNAPSHOTS.md]: done

File: driver/csiplugin/controllerserver.go (hunk 5 above)
Comment (smitaraut) [reply driver/csiplugin/controllerserver.go]: done
